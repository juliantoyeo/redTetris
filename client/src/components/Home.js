import React, { useState, useContext, useEffect } from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import { useHistory } from 'react-router-dom';

import { createPlayer } from '../actions/playerActions';
import { SOCKET_EVENTS } from '../constants/socketConstants';
import { errorAlert } from '../utils/errorUtils';
import { doRoomSocketEvent } from '../utils/socketUtils';

import combinedContext from '../contexts/combinedContext';

import RoomDisplay from './RoomDisplay';
import NameForm from './form/NameForm';
import RoomCreationForm from './form/RoomCreationForm';


const styles = {
	mainContainer: {
		boxSizing: 'border-box',
		display: 'flex',
		alignItems: 'center',
		justifyContent: 'center',
		width: '100vw',
		height: '100vh',
		color: 'white',
		backgroundColor: '#555',
		backgroundSize: 'cover',
		fontFamily: 'Avenir Next',
		fontSize: '1vw'
	},
	box: {
		display: 'flex',
		flexDirection: 'column',
		alignItems: 'center',
		justifyContent: 'center',
		border: '2px solid #333',
		borderRadius: '3vw',
		width: '80vw',
		padding: '2vw'
	},
	roomContainer: {
		width: '100%',
		minHeight: '2vw',
		textAlign: 'center',
		padding: '1vw 0',
		maxHeight: '20vw',
		overflow: 'auto'
	}
}

const Home = (props) => {
	const { socket } = props;
	const history = useHistory();
	const [state, dispatch] = useContext(combinedContext);
	const { currentPlayer, rooms } = state;
	const [form, setForm] = useState({ playerName: '', roomName: '', maxPlayer: 10 });
	const [selectedRoom, setSelectedRoom] = useState(null);

	useEffect(() => {
		if (socket) {
			socket.off(SOCKET_EVENTS.START_GAME);
			socket.on(SOCKET_EVENTS.START_GAME, (updatedRoom) => {
				console.log('WebSocket start game event received :', updatedRoom);
				history.push(`/${updatedRoom.name}[${currentPlayer.name}]`);
			});
		}
	}, [socket, currentPlayer]);

	const onFormChange = (event, type) => {
		const value = event.target.value;
		setForm(prev => ({
			playerName: type === 'playerName' ? value : prev.playerName,
			roomName: type === 'roomName' ? value : prev.roomName,
			maxPlayer: type === 'maxPlayer' ? parseInt(value) : prev.maxPlayer
		}))
	}

	const onSubmitName = async (event) => {
		event.preventDefault();
		createPlayer(socket, form.playerName)(dispatch);
	}

	const onCreateRoom = (event) => {
		event.preventDefault();
		const data = {
			newRoom: {
				name: form.roomName,
				owner: currentPlayer.name,
				maxPlayer: form.maxPlayer
			},
			playerName: currentPlayer.name,
			isJoinRoom: true
		}
		doRoomSocketEvent(socket, setSelectedRoom, SOCKET_EVENTS.CREATE_ROOM, data);
	}

	const onJoinOrLeaveRoom = (room, isJoinRoom) => {
		const data = {
			roomName: room.name,
			playerName: currentPlayer.name,
			isJoinRoom
		}
		if (isJoinRoom || room.players.length > 1) {
			doRoomSocketEvent(socket, setSelectedRoom, SOCKET_EVENTS.UPDATE_ROOM, data);
		} else {
			doRoomSocketEvent(socket, setSelectedRoom, SOCKET_EVENTS.DELETE_ROOM, data);
		}
	}

	const onStartGame = (room) => {
		if (socket)
			socket.emit(SOCKET_EVENTS.START_GAME, room.name, (res) => {
				if (res.status !== 200) {
					errorAlert(res.msg);
				}
			});
	}

	const getRooms = () => {
		if (selectedRoom) {
			const room = _.find(rooms, (room) => room.name === selectedRoom.name);
			if (room) return <RoomDisplay room={room} isLobby={true} isOwner={room.owner === currentPlayer.name} onClick={onStartGame} onLeave={(room) => onJoinOrLeaveRoom(room, false)} />;
		}
		else if (!_.isEmpty(rooms))
			return _.map(rooms, (room, index) => <RoomDisplay room={room} onClick={(room) => onJoinOrLeaveRoom(room, true)} key={`${room.name}#${index}`} />);
		return <RoomDisplay room={null} onClick={null} />;
	}

	return (
		<div style={styles.mainContainer}>
			{!_.isEmpty(currentPlayer) ?
				<div style={styles.box}>
					<span className={'header'}>{!selectedRoom ? 'Available Rooms' : 'Game Lobby'}</span>
					<div style={styles.roomContainer}>
						{getRooms()}
					</div>
					{!selectedRoom && <RoomCreationForm onSubmit={onCreateRoom} onChange={onFormChange} />}
				</div> :
				<div style={styles.box}>
					<NameForm onSubmit={onSubmitName} onChange={onFormChange} />
				</div>}
		</div>
	);
}

Home.propTypes = {
	socket: PropTypes.object
};

export default Home;