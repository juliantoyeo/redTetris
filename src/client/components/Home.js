import React, { useState, useEffect } from 'react'
import { COLORS } from '../constants/gameConstant'

const styles = {
	mainContainer: {
		boxSizing: 'border-box',
		display: 'flex',
		alignItems: 'center',
		justifyContent: 'center',
		width: '100vw',
		height: '100vh',
		color: 'white',
		backgroundColor: '#555',
		backgroundSize: 'cover',
		fontFamily: "Avenir Next",
		fontSize: '1.2vw',
	},
	box: {
		display: 'flex',
		flexDirection: 'column',
		alignItems: 'center',
		justifyContent: 'center',
		border: '2px solid #333',
		borderRadius: '3vw',
		width: '50vw',
		padding: '2vw'
		// minheight: '10vw'
	},
	form: {
		// marginTop: "50px",
		width: "100%",
		display: 'flex',
		flexDirection: 'column',
		alignItems: 'center',
		justifyContent: 'space-evenly',
		textAlign: 'center',
		// padding: '2vw',
		height: '15vw',
		// border: '2px solid #333'
	},
	inputLabel: {
		// fontSize: "1em",
		width: window.innerWidth <= 600 ? "100%" : "50%",
		// marginBottom: "10px",
		// marginRight: "10px",
	},
	textInput: {
		display: "flex",
		flexDirection: "row",
		// fontSize: "1em",
		border: "1px solid #ccc",
		color: "#6A6A6A",
		width: "100%",
		height: "3vw",
		textAlign: 'center',
		// padding: "0.5vw",
		justifyContent: "space-between",
		// marginBottom: "10px",
		// marginRight: "10px",
		backgroundColor: "white",
	},
	submit: {
		backgroundColor: `rgba(${COLORS.RED}, 1)`,
		borderRadius: "1vw",
		fontSize: "1em",
		height: "3vw",
		width: window.innerWidth <= 600 ? "100%" : "20%",
		// marginTop: "30px",
		color: "white",
		// padding: "1vw",
		cursor: "pointer",
	},
}

const Home = () => {
	const [form, setform] = useState({})

	const onSearchChange = (event, type) => {
		if (type == "playerName") {
			const playerName = _.toUpper(event.target.value)
			setform({
				...form,
				playerName,
			})
		}
	}

	const onSubmit = async (event) => {
		event.preventDefault()
	}

	return (
		<div style={styles.mainContainer}>
			<div style={styles.box}>
				Welcome to Red Tetris!
				<form style={styles.form} onSubmit={onSubmit}>
					<div style={styles.inputLabel}>
						Enter Your Name
					</div>
					<label style={styles.inputLabel}>
						<input
							name={"playerName"}
							style={{ ...styles.textInput, ...styles.upperCase}}
							type={"text"}
							defaultValue={form.playerName}
							onChange={(e) => onSearchChange(e, "playerName")}
							required
						/>
					</label>
					<button style={styles.submit} type={"submit"}>Play</button>
				</form>
			</div>
		</div>
	)
}

export default Home