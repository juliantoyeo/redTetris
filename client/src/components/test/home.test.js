import React from 'react';
import { shallow } from 'enzyme';
import enableHooks from 'jest-react-hooks-shallow';
import _ from 'lodash';

import Home from '../Home';
import * as AppContext from '../../contexts/combinedContext';

// import Client from 'socket.io-client';
// import params from '../../params';
// let socket = new Client(params.server.url);

import MockedSocket from 'socket.io-mock';
let socket = new MockedSocket();
socket.emit = jest.fn();

enableHooks(jest);
// import RoomDisplay from '../RoomDisplay'
// jest.mock('../RoomDisplay')
// eslint-disable-next-line react/display-name
// RoomDisplay.mockImplementation = () => <div />

let state = {
	currentPlayer: {
		name: 'player1'
	},
	rooms: [
		{
			name: 'room1'
		}
	]
}

let fakeRoom = {
	name: 'fakeRoom',
	players: ['fakePlayer1', 'fakePlayer2']
}


describe('Test Home component', () => {

	const setState = jest.fn();

	beforeEach(() => {

		const contextValues = [state, jest.fn()];
		jest
			.spyOn(AppContext, 'useAppContext')
			.mockImplementation(() => contextValues);
		
		jest.spyOn(React, 'useEffect').mockImplementation(f => f());

		React.useState = jest.fn()
			.mockReturnValueOnce([null, setState])
			.mockReturnValueOnce([{ playerName: '', roomName: '', maxPlayer: 10 }, setState])
	});

	afterEach(() => {
		jest.clearAllMocks();
	});

	it('Should render properly', () => {

		const wrapper = shallow(<Home />);
		const roomDisplay = wrapper.find('RoomDisplay');

		roomDisplay.props().onClick(fakeRoom, true);
		expect(wrapper).toMatchSnapshot();
	});
	it('Should render selected room and test onStartGame', () => {

		React.useState = jest.fn()
			.mockReturnValueOnce([{ name: 'room1' }, setState])
			.mockReturnValueOnce([{ playerName: '', roomName: '', maxPlayer: 10 }, setState])
		const wrapper = shallow(<Home socket={socket} />);
		const roomDisplay = wrapper.find('RoomDisplay');
		// socket.emit.mockResolvedValue({ status: '400' });
		roomDisplay.simulate('click', { name: 'test' });
		
		roomDisplay.props().onLeave(fakeRoom, true);
		roomDisplay.props().onLeave({ ...fakeRoom, players: [] }, false);
		expect(wrapper).toMatchSnapshot();
	});
	it('Should test onStartGame with null socket', () => {

		React.useState = jest.fn()
			.mockReturnValueOnce([{ name: 'room1' }, setState])
			.mockReturnValueOnce([{ playerName: '', roomName: '', maxPlayer: 10 }, setState])

		const wrapper = shallow(<Home />);
		const roomDisplay = wrapper.find('RoomDisplay');
		roomDisplay.simulate('click', { name: 'test' });
	});
	it('Should render no room found', () => {

		React.useState = jest.fn()
			.mockReturnValueOnce([{ name: 'non existing room' }, setState])
			.mockReturnValueOnce([{ playerName: '', roomName: '', maxPlayer: 10 }, setState])

		const wrapper = shallow(<Home />);
		expect(wrapper).toMatchSnapshot();
	});
	it('Should render no room', () => {
		let newState = _.cloneDeep(state);
		newState.rooms = [];
		const contextValues = [newState, jest.fn()];
		jest
			.spyOn(AppContext, 'useAppContext')
			.mockImplementation(() => contextValues);

		const wrapper = shallow(<Home />);
		expect(wrapper).toMatchSnapshot();
	});
	test('Should test onFormChange, onCreateRoom', () => {

		const wrapper = shallow(<Home />);
		const roomCreationForm = wrapper.find('RoomCreationForm');
		roomCreationForm.simulate('change', { target: { value: 'test' } });
		roomCreationForm.simulate('submit', { preventDefault: jest.fn() });
	});
	test('Should test onSubmitName', () => {
		let newState = _.cloneDeep(state);
		newState.currentPlayer = null;
		const contextValues = [newState, jest.fn()];
		jest
			.spyOn(AppContext, 'useAppContext')
			.mockImplementation(() => contextValues);

		const wrapper = shallow(<Home />);
		const nameForm = wrapper.find('NameForm');
		nameForm.simulate('submit', { preventDefault: jest.fn() });
	});
});