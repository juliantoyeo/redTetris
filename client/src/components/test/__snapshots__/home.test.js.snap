// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Home component Should render no room 1`] = `
<div
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#555",
      "backgroundSize": "cover",
      "boxSizing": "border-box",
      "color": "white",
      "display": "flex",
      "fontFamily": "Avenir Next",
      "fontSize": "1vw",
      "height": "100vh",
      "justifyContent": "center",
      "width": "100vw",
    }
  }
>
  <div
    style={
      Object {
        "alignItems": "center",
        "border": "2px solid #333",
        "borderRadius": "3vw",
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "center",
        "padding": "2vw",
        "width": "80vw",
      }
    }
  >
    <span
      className="header"
    >
      Available Rooms
    </span>
    <div
      style={
        Object {
          "maxHeight": "20vw",
          "minHeight": "2vw",
          "overflow": "auto",
          "padding": "1vw 0",
          "textAlign": "center",
          "width": "100%",
        }
      }
    >
      <RoomDisplay
        onClick={null}
        room={null}
      />
    </div>
    <RoomCreationForm
      onChange={[Function]}
      onSubmit={[Function]}
    />
  </div>
</div>
`;

exports[`Test Home component Should render no room found 1`] = `
<div
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#555",
      "backgroundSize": "cover",
      "boxSizing": "border-box",
      "color": "white",
      "display": "flex",
      "fontFamily": "Avenir Next",
      "fontSize": "1vw",
      "height": "100vh",
      "justifyContent": "center",
      "width": "100vw",
    }
  }
>
  <div
    style={
      Object {
        "alignItems": "center",
        "border": "2px solid #333",
        "borderRadius": "3vw",
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "center",
        "padding": "2vw",
        "width": "80vw",
      }
    }
  >
    <span
      className="header"
    >
      Game Lobby
    </span>
    <div
      style={
        Object {
          "maxHeight": "20vw",
          "minHeight": "2vw",
          "overflow": "auto",
          "padding": "1vw 0",
          "textAlign": "center",
          "width": "100%",
        }
      }
    >
      <RoomDisplay
        onClick={null}
        room={null}
      />
    </div>
  </div>
</div>
`;

exports[`Test Home component Should render properly 1`] = `
<div
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#555",
      "backgroundSize": "cover",
      "boxSizing": "border-box",
      "color": "white",
      "display": "flex",
      "fontFamily": "Avenir Next",
      "fontSize": "1vw",
      "height": "100vh",
      "justifyContent": "center",
      "width": "100vw",
    }
  }
>
  <div
    style={
      Object {
        "alignItems": "center",
        "border": "2px solid #333",
        "borderRadius": "3vw",
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "center",
        "padding": "2vw",
        "width": "80vw",
      }
    }
  >
    <span
      className="header"
    >
      Available Rooms
    </span>
    <div
      style={
        Object {
          "maxHeight": "20vw",
          "minHeight": "2vw",
          "overflow": "auto",
          "padding": "1vw 0",
          "textAlign": "center",
          "width": "100%",
        }
      }
    >
      <RoomDisplay
        key="room1#0"
        onClick={[Function]}
        room={
          Object {
            "name": "room1",
          }
        }
      />
    </div>
    <RoomCreationForm
      onChange={[Function]}
      onSubmit={[Function]}
    />
  </div>
</div>
`;

exports[`Test Home component Should render selected room and test onStartGame 1`] = `
<div
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#555",
      "backgroundSize": "cover",
      "boxSizing": "border-box",
      "color": "white",
      "display": "flex",
      "fontFamily": "Avenir Next",
      "fontSize": "1vw",
      "height": "100vh",
      "justifyContent": "center",
      "width": "100vw",
    }
  }
>
  <div
    style={
      Object {
        "alignItems": "center",
        "border": "2px solid #333",
        "borderRadius": "3vw",
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "center",
        "padding": "2vw",
        "width": "80vw",
      }
    }
  >
    <span
      className="header"
    >
      Game Lobby
    </span>
    <div
      style={
        Object {
          "maxHeight": "20vw",
          "minHeight": "2vw",
          "overflow": "auto",
          "padding": "1vw 0",
          "textAlign": "center",
          "width": "100%",
        }
      }
    >
      <RoomDisplay
        isLobby={true}
        isOwner={false}
        onClick={[Function]}
        onLeave={[Function]}
        room={
          Object {
            "name": "room1",
          }
        }
      />
    </div>
  </div>
</div>
`;
